rust/lance/src/io/exec/filtered_read.rs

```diff
@@ -520,158 +470,291 @@ impl FilteredReadStream {
     // We need to figure out which ranges to read from each fragment.
     //
     // If the scan range is ignoring the filters we can push it down here.
-    // If the scan range is not ignoring the filters we can only push it down if the index
-    // result is an exact match.
+    // If the scan range is not ignoring the filters we can only push it down if:
+    // 1. The index result is an exact match (we know exactly which rows will be in the result)
+    // 2. The index result is AtLeast with guaranteed rows >= limit (we have enough guaranteed matches)
+    // Returns: (fragment reads, whether limit was pushed down to fragment ranges)
     #[instrument(name = "plan_scan", skip_all)]
     async fn plan_scan(
         dataset: &Dataset,
         fragments: Vec<LoadedFragment>,
         evaluated_index: &Option<Arc<EvaluatedIndex>>,
-        options: FilteredReadOptions,
+        options: &FilteredReadOptions,
         scan_scheduler: Arc<ScanScheduler>,
-    ) -> Result<Vec<ScopedFragmentRead>> {
-        let mut scoped_fragments = Vec::with_capacity(fragments.len());
-        let projection = Arc::new(options.projection);
+    ) -> Result<(Vec<ScopedFragmentRead>, bool)> {
+        // For pushing down scan_range_after_filter
+        let mut scan_planned_with_limit_pushed_down = false;
+        let mut to_skip = options
+            .scan_range_after_filter
+            .as_ref()
+            .map(|r| r.start)
+            .unwrap_or(0);
+        let mut to_take = options
+            .scan_range_after_filter
+            .as_ref()
+            .map(|r| r.end - r.start)
+            .unwrap_or(u64::MAX);
+
+        let mut fragments_to_read: HashMap<u32, Vec<Range<u64>>> = HashMap::new();
+        let mut scan_push_down_fragments_to_read: HashMap<u32, Vec<Range<u64>>> = HashMap::new();
```

Comment:
I'm not sure I understand the purpose of this variable.  Can you add a comment here explaining why we need two maps?